File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def differ(self, root, minVal, maxVal):
        if not root:
            return maxVal - minVal
        
        minVal = min(root.val, minVal)
        maxVal = max(root.val, maxVal)
        
        leftDiff = self.differ(root.left, minVal, maxVal)
        rightDiff = self.differ(root.right, minVal, maxVal)
        
        return max(leftDiff, rightDiff)
    
    def maxAncestorDiff(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        return self.differ(root, root.val, root.val)

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        if (!root) {
            return 0;
        }
        
        return differ(root, root->val, root->val);
    }

//helper function to calculate largest difference
private:
    int differ(TreeNode* root, int minVal, int maxVal) {
        if (!root) {
            return maxVal - minVal;
        }
        
        //compares minimum and maximum values at each traversal
        minVal = min(minVal, root->val);
        maxVal = max(maxVal, root->val);

        int leftDiff = differ(root->left, minVal, maxVal);
        int rightDiff = differ(root->right, minVal, maxVal);

        return max(leftDiff, rightDiff);
    }
};